### PR TITLE
Protobuf generate V2 and V3 protos and extend tests. 

### PR DESCRIPTION
**Pull request by rangadi**
This is a temporary PR on, made on top of Java-support PR #37738. 

Outline:
  * Generates descriptor files and java classes for both V2 and V3
  * Updates the unit tests to check all variations (V2 & V3 x Java, decriptor file).
  * Adds a test with default values and required fields for V2.
  * Maven build is updated. But not clear how do the same in SBT.

### CODE REVIEW

**Comment by rangadi** [comment]
FYI: @mposdev21, @SandishKumarHN

**Review by mposdev21 — COMMENTED** [review]
Overall, looks good to me except for a few things I noted.

**Comment by LuciferYang** [comment]
A newbie question: why we need support protobuf v2?

File: connector/protobuf/src/main/scala/org/apache/spark/sql/protobuf/ProtobufDataToCatalyst.scala [hunk 1]
```diff
@@ -55,10 +55,14 @@ private[protobuf] case class ProtobufDataToCatalyst(
   private lazy val protobufOptions = ProtobufOptions(options)
 
   @transient private lazy val messageDescriptor =
-    ProtobufUtils.buildDescriptor(descFilePath, messageName)
+    ProtobufUtils.buildDescriptor(messageName, descFilePath)
+    // TODO: Avoid carrying the file name. Read the contents of descriptor file only once
+    //       at the start. Rest of the runs should reuse the buffer. Otherwise, it could
+    //       cause inconsistencies if the file contents are changed the user after a few days.
+    //       Same for the write side in [[CatalystDataToProtobuf]].
```
[the remark this answers is not in the document]
Comment (rangadi) [reply connector/protobuf/src/main/scala/org/apache/spark/sql/protobuf/ProtobufDataToCatalyst.scala]: I had added this in my Java class pr #38286. 
The file contents are read on executors too when the `ProtobufDataToCatalyst()` class is initialized there. Notice `@transient private lazy val messageDescriptor` in the class.

File: connector/protobuf/src/test/protobuf/v2/catalyst_types.proto [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+// protoc --java_out=connector/protobuf/src/test/resources/protobuf/ connector/protobuf/src/test/resources/protobuf/catalyst_types.proto
+// protoc --descriptor_set_out=connector/protobuf/src/test/resources/protobuf/catalyst_types.desc --java_out=connector/protobuf/src/test/resources/protobuf/org/apache/spark/sql/protobuf/ connector/protobuf/src/test/resources/protobuf/catalyst_types.proto
+
```
[the remark this answers is not in the document]
Comment (rangadi) [reply connector/protobuf/src/test/protobuf/v2/catalyst_types.proto:17]: Removed. Also added `pyspark_test.proto` to this directory.

File: connector/protobuf/pom.xml [hunk 3]
```diff
@@ -110,6 +109,60 @@
...
+                <outputTarget>
+                  <type>descriptor</type>
+                  <addSources>test</addSources>
+                  <outputDirectorySuffix>descriptor-set-v3</outputDirectorySuffix>
+                </outputTarget>
+              </outputTargets>
+            </configuration>
+          </execution>
+        </executions>
+      </plugin>
```
[the remark this answers is not in the document]
Comment (rangadi) [reply connector/protobuf/pom.xml]: Fixed now.

File: connector/protobuf/src/test/scala/org/apache/spark/sql/protobuf/ProtobufFunctionsSuite.scala [hunk 4]
```diff
@@ -56,49 +103,50 @@ class ProtobufFunctionsSuite extends QueryTest with SharedSparkSession with Seri
...
+        from_protobuf_wrapper($"value_to", name, descFilePathOpt).as("value_to_from"))
+      checkAnswer(fromProtoDF.select($"value_from.*"), toFromProtoDF.select($"value_to_from.*"))
+    }
   }
 
   test("roundtrip in from_protobuf and to_protobuf - Repeated Message Once") {
-    val repeatedMessageDesc = ProtobufUtils.buildDescriptor(testFileDesc, "RepeatedMessage")
-    val basicMessageDesc = ProtobufUtils.buildDescriptor(testFileDesc, "BasicMessage")
+    val repeatedMessageDesc = ProtobufUtils.buildDescriptor(descPathV3, "RepeatedMessage")
+    val basicMessageDesc = ProtobufUtils.buildDescriptor(descPathV3, "BasicMessage")
```
[the remark this answers is not in the document]
Comment (rangadi) [reply connector/protobuf/src/test/scala/org/apache/spark/sql/protobuf/ProtobufFunctionsSuite.scala:149]: Yes.

**Comment by rangadi** [comment]
> A newbie question: why we need support protobuf v2?

@LuciferYang many users still use v2 protobufs. It is not always easy to migrate the protobufs to v3 or even necessary. Note that v2 protobufs are still genrated with `protoc` 3.x.x. The v2 protos used in tests are built with 3.2.1. The syntax version (v2 or v3) is different from version used to parse and generate Java classes or descriptor sets.

**Pull request closed without merging**